types/wlr_xdg_shell_v6.c

```diff
@@ -91,19 +120,40 @@ static const struct zxdg_toplevel_v6_interface zxdg_toplevel_v6_implementation =
 	.set_minimized = xdg_toplevel_set_minimized
 };
 
-static void xdg_surface_destroy(struct wl_resource *resource) {
-	struct wlr_xdg_surface_v6 *surface = wl_resource_get_user_data(resource);
+static void xdg_surface_destroy(struct wlr_xdg_surface_v6 *surface) {
+	wl_resource_set_user_data(surface->resource, NULL);
 	wl_list_remove(&surface->link);
+	wl_list_remove(&surface->surface_destroy_listener.link);
+	wl_list_remove(&surface->surface_commit_listener.link);
 	free(surface);
 }
 
+static void xdg_surface_resource_destroy(struct wl_resource *resource) {
+	struct wlr_xdg_surface_v6 *surface = wl_resource_get_user_data(resource);
+	if (surface != NULL) {
+		xdg_surface_destroy(surface);
+	}
+}
+
+static const char* wlr_desktop_xdg_toplevel_role = "xdg_toplevel";
+
 static void xdg_surface_get_toplevel(struct wl_client *client,
 		struct wl_resource *resource, uint32_t id) {
 	// TODO: Flesh out
```

Comment:
Is this TODO still valid?

Reply:
I think the only thing left to do here is initialize the toplevel state.